### PR TITLE
Properly retain types of Mapping subclasses

### LINKED ISSUE #1536: defaultdict converted to dict after __post_init__
# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.5.1
            pydantic compiled: True
                 install path: <redacted>/lib/python3.7/site-packages/pydantic
               python version: 3.7.3 | packaged by conda-forge | (default, Dec  6 2019, 08:36:57)  [Clang 9.0.0 (tags/RELEASE_900/final)]
                     platform: Darwin-19.4.0-x86_64-i386-64bit
     optional deps. installed: ['typing-extensions']
```

If I run the following snippet:
```py
import collections
import dataclasses
from typing import DefaultDict

from pydantic.dataclasses import dataclass


@dataclass
class X:
    f: DefaultDict[int, int] = dataclasses.field(init=False)

    def __post_init__(self):
        self.f = collections.defaultdict(int)
        self.f[2] += 1
        print(f"__post_init__: {self.f}")


x = X()
print(type(x.f))
print(x.f)
```

It prints the following:
```
__post_init__: defaultdict(<class 'int'>, {2: 1})
<class 'dict'>
{2: 1}
```

In other words, the `defaultdict` field is converted into a `dict` (while preserving the keys), even though the field is annotated as DefaultDict, which is surprising. Is this intended behavior?

### LINKED ISSUE #2323: Mapping types are coerced to dict
# Bug

Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.5.1
            pydantic compiled: True
                 install path: C:\Users\ofek\AppData\Local\Programs\Python\Python38\Lib\site-packages\pydantic
               python version: 3.8.1 (tags/v3.8.1:1b293b6, Dec 18 2019, 23:11:46) [MSC v.1916 64 bit (AMD64)]
                     platform: Windows-10-10.0.19041-SP0
     optional deps. installed: ['typing-extensions', 'email-validator']
```

I'm trying to use https://github.com/MagicStack/immutables

```py
from __future__ import annotations

from typing import Mapping

from immutables import Map
from pydantic import BaseModel


class Model(BaseModel):
    test: Mapping[str, int]

print(type(Model(test=Map(key=42)).test))
```

produces

```
<class 'dict'>
```

Is there a way to keep the type of encountered Mappings, or create a type representing `immutables.Map`?

I've been at this for hours reading docs/closed issues, trying various approaches, and now I'm simply at a loss here.

### PR DESCRIPTION
**Pull request by ofek**
## Change Summary

This change prevents Mapping subclasses from always being coerced to `dict`

## Related issue number

Resolves #2323 
closes #1536

## Checklist

* [X] Unit tests for the changes exist
* [X] Tests pass on CI and coverage remains at 100%
* [X] Documentation reflects the changes where applicable
* [X] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by ofek** [comment]
@PrettyWood I'm not sure how to fix the type checking errors

**Comment by ofek** [comment]
@PrettyWood The current job failure is due to https://pypi.org/project/cryptography/3.4.1/ emitting a warning in the test suite of FastAPI

**Comment by PrettyWood** [comment]
Yep can't do anything about that now but it should be handled rapidly on FastAPI side.
In the meantime you can add some tests to keep coverage at 100% :)

**Comment by ofek** [comment]
Done!

**Review by PrettyWood — CHANGES_REQUESTED** [review]
@ofek We should still check the set type and try to coerce when needed.
For example
```py
class M(BaseModel):
    x: Dict[str, int]
```
should coerce to `dict`. Right now there is no difference between `Dict` and `Mapping`...
I updated https://github.com/PrettyWood/pydantic/pull/70/files if you want

-----------------------------------------------------------------

Note that I think we should probably have a proper validator for `Dict` (at least) before `Mapping` to avoid any performance impact on normal dictionaries (like it's done for `Tuple` or `List`, which are treated before `Sequence`)

**Comment by ofek** [comment]
Updated based on your branch

File: pydantic/fields.py [hunk 1]
```diff
@@ -737,7 +738,25 @@ def _validate_mapping(
...
+            return self._get_mapping_value(v, result), None
+
+    def _get_mapping_value(self, original: T, converted: Dict[Any, Any]) -> Union[T, Dict[Any, Any]]:
+        target_type = get_origin(self.outer_type_)
+        original_type = type(original)
+
+        if is_mapping_type(target_type):
+            target_type = original_type
+
+        if target_type is dict:
```
Comment (PrettyWood) [inline pydantic/fields.py:750]:
```suggestion
        if target_type in {dict, Dict}:
```

Comment (PrettyWood) [reply pydantic/fields.py:750]: For python 3.6, `__origin__` is `typing.Dict`, which is just an alias for `dict`

File: pydantic/fields.py [hunk 2]
```diff
@@ -737,7 +738,25 @@ def _validate_mapping(
...
+            return converted
+        elif target_type is defaultdict:
```
Comment (PrettyWood) [inline pydantic/fields.py:752]:
```suggestion
        elif target_type in {defaultdict, DefaultDict}:
```
same

File: pydantic/fields.py [hunk 3]
```diff
@@ -737,7 +738,25 @@ def _validate_mapping(
...
+            return defaultdict(getattr(original, 'default_factory', None), **converted)
```
Comment (PrettyWood) [inline pydantic/fields.py:753]:
```suggestion
            return defaultdict(getattr(original, 'default_factory', None), converted)
```
and same under.
If we use `kwargs`, it won't work for `int` keys. But it does when passing a dict as `arg`.
Same under

**Review by PrettyWood — COMMENTED** [review]
I still reckon we should at least have `SHAPE_DICT` (extra could be `SHAPE_DEFAULTDICT`, `SHAPE_COUNTER`...) before `SHAPE_MAPPING` to have no perf impact.
Or at least if we go with it we should have a benchmark to see the impact

**Comment by ofek** [comment]
@PrettyWood Addressed. I'm not familiar with the concept of shapes here. Can you explain?

**Comment by PrettyWood** [comment]
@ofek _pydantic_ is fast because a lot is done in the metaclass. You can have a look at `ModelField._type_analysis()`, which is run in `ModelMetaclass` when `ModelField.infer()` is called.
In `_type_analysis` a lot is done to set the right attributes to each ModelField (type, shape, validators...)
And the `shape` is used to call the right default validators to speed up the parsing

**Comment by ofek** [comment]
I just tried that but many tests are failing now

**Comment by ofek** [comment]
should I revert the last commit for now?

**Comment by ofek** [comment]
I'm still trying but I don't understand what's wrong 🙁

**Comment by ofek** [comment]
@PrettyWood Could you please take a look when you get a chance?

**Comment by PrettyWood** [comment]
I'll have a look tonight or tomorrow 👍

**Comment by ofek** [comment]
Thank you, feel free to push directly to my branch!

**Comment by PrettyWood** [comment]
Hi @ofek
Had a quick look and got all tests working in local https://github.com/PrettyWood/pydantic/pull/70
I have some guests coming around soon but I think you should be able to work it out without a problem
I think you still need to add some tests to showcase a random mapping getting coerced into `dict` when `Dict[...]` is set.
Hope it helps

**Comment by ofek** [comment]
@PrettyWood Done! Added that test too

File: tests/test_main.py [hunk 4]
```diff
@@ -1472,3 +1473,71 @@ class Item(BaseModel):
...
     assert id(image_1) == id(item.images[0])
     assert id(image_2) == id(item.images[1])
+
+
+def test_mapping_retains_type_subclass():
+    class Map(dict):
+        pass
+
+    class Model(BaseModel):
+        field: Mapping[str, Mapping[str, int]]
```
Comment (samuelcolvin) [inline tests/test_main.py:1483]: this is a very confusing name for a field.

File: tests/test_main.py [hunk 5]
```diff
@@ -1472,3 +1473,71 @@ class Item(BaseModel):
 
     assert id(image_1) == id(item.images[0])
     assert id(image_2) == id(item.images[1])
+
+
+def test_mapping_retains_type_subclass():
+    class Map(dict):
```
Comment (samuelcolvin) [inline tests/test_main.py:1479]:
```suggestion
    class CustomMap(dict):
```
might be a clearer name?

**Review by samuelcolvin — COMMENTED** [review]
otherwise this is looking good.

**Comment by ofek** [comment]
@samuelcolvin Thanks, addressed!

**Comment by ofek** [comment]
Done this look alright now?

**Comment by ofek** [comment]
@samuelcolvin Just let me know if I have to do anything else! FYI, after this is merged and released, every [Datadog Agent integration](https://github.com/DataDog/integrations-core) will exclusively use pydantic for user configuration 🙂

**Comment by samuelcolvin** [comment]
This is great. Thank you.

> @samuelcolvin Just let me know if I have to do anything else! FYI, after this is merged and released, every [Datadog Agent integration](https://github.com/DataDog/integrations-core) will exclusively use pydantic for user configuration 🙂

I don't see pydantic as a dependency, what am I missing?

**Comment by PrettyWood** [comment]
@samuelcolvin https://github.com/DataDog/integrations-core/pull/8675

**Comment by ofek** [comment]
https://github.com/DataDog/integrations-core/pull/8675 is merged and I've started the syncing of every integration: https://github.com/DataDog/integrations-core/pulls 😄

**Comment by ofek** [comment]
https://datadoghq.dev/integrations-core/meta/config-models/

**Pull request merged**